### ISSUE TITLE
Followup Issue for #8654 Enhanced depends logic not working when task is omitted

### ISSUE THREAD

**Issue by mhehle** [argoproj/argo-workflows#9153]
## Checklist

 

* [x] Double-checked my configuration.
* [x] Tested using the latest version.
* [x] Used the Emissary executor.

## Summary
This is a follow up Issue for #8654 . This issue still exists in version 3.3.8 . 
Its impossible to launch/create a workflow with `depends` on `.Omitted` step.
You get an error: `Internal Server Error: templates.main.tasks.E task result 'Omitted' for task 'D' is invalid`

**What happened/what you expected to happen?**
When a task is Skipped/Omitted because a dependent task is skipped you can't use enhanced depends on that task.

```
  O
 /  \
A    B <- B is skipped
|    |
C    D <- D gets omitted
|   /
E <- if C.Succeeded && (D.Succeeded || D.Omitted)*
```

*doesn't work

**What version are you running?**
3.3.8

## Diagnostics

Paste the smallest workflow that reproduces the bug. We must be able to run the workflow.

```yaml
apiVersion: argoproj.io/v1alpha1
metadata:
  name: broken-depends
  generateName: broken-depends-
  namespace: argo
spec:
  serviceAccountName: workflow
  entrypoint: main
  templates:
    - name: main
      inputs:
        parameters:
          - name: input_var
            default: 'true'
      dag:
        tasks:
          - name: A
            template: pass
          - name: B
            template: pass
            when: '{{ inputs.parameters.input_var }} == false'
          - name: C
            template: pass
            depends: A.Succeeded
          - name: D
            template: pass
            depends: B.Succeeded
          - name: E
            template: pass
            depends: (D.Succeeded || D.Omitted) && C
    - name: pass
      container:
        image: 'alpine:3.7'
        command:
          - sh
          - '-c'
          - exit 0
    - name: fail
      container:
        image: 'alpine:3.7'
        command:
          - sh
          - '-c'
          - exit 1
```

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by mhehle** [comment]
still relevant

**Comment by ezk84** [comment]
This is not completed and still an issue.

**Comment by mhehle** [comment]
the commit [961f731](https://github.com/argoproj/argo-workflows/commit/961f731b7e9cb60490dd763a394893154c0b3c60) with the fix was only picked for 3.3.9 release. So this issue should be reolved since 3.3.9.

CherryPick list for 3.3.9
https://github.com/argoproj/argo-workflows/issues/9262
